File: src/main/java/com/hazza/LeetCodeByJava/SegmentTwo/No116PopulatingNextRightPointersInEachNode.java
package com.hazza.LeetCodeByJava.SegmentTwo;

import com.hazza.LeetCodeByJava.datastructure.TreeLinkNode;

import java.util.ArrayDeque;
import java.util.Queue;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/populating-next-right-pointers-in-each-node/description/
 * User: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 18-1-28
 * Time: 8:30 PM
 */
public class No116PopulatingNextRightPointersInEachNode {

    // solution 1
    public void connect(TreeLinkNode root) {
        if (root == null) return;
        Queue<TreeLinkNode> queue = new ArrayDeque<>();

        queue.offer(root);
        while (!queue.isEmpty()) {
            int len = queue.size();
            for (int i = 0; i < len; i++) {
                TreeLinkNode node = queue.poll();
                if (node.left != null) {
                    queue.offer(node.left);
                    queue.offer(node.right);
                }
                if (i == len - 1) node.next = null;
                else node.next = queue.peek();
            }
        }
    }

    // solution 2
    public void connect1(TreeLinkNode root) {
        if (root == null) return;
        TreeLinkNode prev = root;
        TreeLinkNode cur = null;
        while (prev.left != null) {
            cur = prev;
            while (cur != null) {
                cur.left.next = cur.right;
                if (cur.next != null) cur.right.next = cur.next.left;
                cur = cur.next;
            }
            prev = prev.left;
        }
    }

    // solution 3
    public void connect2(TreeLinkNode root) {
        if (root == null) return;

        if (root.left != null) {
            root.left.next = root.right;
            if (root.next != null) root.right.next = root.next.left;
        }

        connect2(root.left);
        connect2(root.right);
    }
}

File: src/main/python/leetcode_by_python/tree/No116_Populating_Next_Right_Pointers_in_Each_Node.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019-11-03
from src.main.python.leetcode_by_python.data_structure.Node import Node


class Solution:
    def connect1(self, root: Node) -> Node:
        if not root:
            return None
        queue = [root]

        while queue:
            l = len(queue)
            for i in range(l):
                node = queue.pop(0)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
                node.next = None if i == l -1 else queue[0]

        return root

    def connect2(self, root: Node) -> Node:
        cur = root
        while cur and cur.left:
            next = cur.left
            while cur:
                cur.left.next = cur.right
                cur.right.next = cur.next and cur.next.left
                cur = cur.next
            cur = next

        return root

    def connect3(self, root: Node) -> Node:
        def dfs(node):
            if not node:
                return
            if node.left:
                node.left.next = node.right
                if node.next:
                    node.right.next = node.next.left
            dfs(node.left)
            dfs(node.right)

        dfs(root)
        return root

